src/coreclr/jit/flowgraph.cpp

```diff
@@ -4184,3 +4189,298 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
         lvaSetVarDoNotEnregister(lclNum DEBUGARG(DoNotEnregisterReason::LocalField));
     }
 }
+
+//------------------------------------------------------------------------------
+// fgExpandThreadLocalAccess : Expand the CORINFO_HELP_GETSHARED_NONGCTHREADSTATIC_BASE_NOCTOR_OPTIMIZED
+//                             that access fields marked with [ThreadLocal].
+//
+// Returns:
+//    PhaseStatus indicating what, if anything, was changed.
+//
+// Notes:
+//    A cache is stored in thread local storage (TLS) of coreclr. It maps the typeIndex (embedded in
+//    the code at the JIT time) to the base of static blocks. This method generates code to
+//    extract the TLS, get the entry at which the cache is stored. Then it checks if the typeIndex of
+//    enclosing type of current field is present in the cache and if yes, extract out that can be directly
+//    accessed at the uses.
+//    If the entry is not present, the helper is called, which would make an entry of current static block
+//    in the cache.
```

Comment:
Would it be useful to show pseudo-code here of the generated code?

The cache design (and JIT-generated pseudo-code) would be worthy of a short MarkDown document in docs\design\features

Reply:
> Would it be useful to show pseudo-code here of the generated code?

I am not sure if it is useful at all. I already have block layout and c-style pseudo code to get an idea of what it is generating. Let me know if you still think generated code will be of any use.

> The cache design (and JIT-generated pseudo-code) would be worthy of a short MarkDown document in docs\design\features

It is an existing mechanism that is already done in NativeAOT. https://github.com/dotnet/runtime/blob/1d2cd206f40306cbf06f3e2676e7cfa301077285/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/ThreadStatics.cs#L24-L87